restli-server/src/main/java/com/linkedin/restli/internal/server/response/BatchFinderResponseBuilder.java

```diff
@@ -195,26 +196,52 @@ private AnyRecord buildMetaData(CollectionResult<RecordTemplate, RecordTemplate>
     return null;
   }
 
-  private String getBatchParamterName(RoutingResult routingResult)
+  private String getBatchParameterName(RoutingResult routingResult)
   {
     int batchFinderCriteriaIndex = routingResult.getResourceMethod().getBatchFinderCriteriaParamIndex();
     return routingResult.getResourceMethod().getParameters().get(batchFinderCriteriaIndex).getName();
   }
 
   private DataList getCriteriaParameters(RoutingResult routingResult)
   {
-    String batchParamName = getBatchParamterName(routingResult);
-    return(DataList)routingResult.getContext().getStructuredParameter(batchParamName);
+    String batchParameterName = getBatchParameterName(routingResult);
+    return(DataList)routingResult.getContext().getStructuredParameter(batchParameterName);
   }
 
-  private URI buildCriteriaURI(ResourceContextImpl resourceContext, RecordTemplate criteria, String queryParam, URI uri)
+  protected static URI buildCriteriaURI(ResourceContextImpl resourceContext, RecordTemplate criteria, String batchParameterName, URI uri)
   {
     DataList criteriaList = new DataList(1);
     criteriaList.add(criteria.data());
+    DataMap queryParams = extractQueryParamsFromResourceContext(resourceContext);
     return URIParamUtils.replaceQueryParam(uri,
-                                           queryParam,
+                                           batchParameterName,
                                            criteriaList,
-                                           resourceContext.getParameters(),
+                                           queryParams,
                                            resourceContext.getRestliProtocolVersion());
   }
+
+  protected static DataMap extractQueryParamsFromResourceContext(ResourceContextImpl resourceContext)
+  {
+    try
+    {
+      DataMap queryParams = resourceContext.getParameters().clone();
+      if (queryParams.containsKey(RestConstants.FIELDS_PARAM))
+      {
+        queryParams.put(RestConstants.FIELDS_PARAM, resourceContext.getProjectionMask().getDataMap());
+      }
+      if (queryParams.containsKey(RestConstants.PAGING_FIELDS_PARAM))
+      {
+        queryParams.put(RestConstants.PAGING_FIELDS_PARAM, resourceContext.getPagingProjectionMask().getDataMap());
+      }
+      if (queryParams.containsKey(RestConstants.METADATA_FIELDS_PARAM))
+      {
+        queryParams.put(RestConstants.METADATA_FIELDS_PARAM, resourceContext.getMetadataProjectionMask().getDataMap());
+      }
+      return queryParams;
+    }
+    catch (CloneNotSupportedException exception)
+    {
+      throw new RuntimeException(exception);
```

Comment:
In this part of error handling, because I don't expect to change all the callers of buildCriteriaURI with `throws`
so I instead throws a RuntimeException

Reply:
I think throwing a RestLiServiceException with 500 status code and cleaner error message would be good here. For reference you can check the other response builders.

Reply:
I just fixed it, one point I am curious is that, for each entry in the batch, is it better to put in 500 error code for each entry, or just fail the whole batch in one 500 code?

Reply:
For this, just one 500 error is fine. The paging parameters are common for all criteria.

Reply:
got it I will make another change then, right now i put failure in one entry, the other will continue to process or they will generate another 500.

Reply:
@karthikbalasub , all the changes regarding clone and exception handling are reverted since I use a simpler way to address the original issue